packages/@aws-cdk/aws-stepfunctions-tasks/lib/send-to-queue.ts

```diff
@@ -48,6 +55,9 @@ export interface SendToQueueProps {
  */
 export class SendToQueue implements sfn.IStepFunctionsTask {
   constructor(private readonly queue: sqs.IQueue, private readonly props: SendToQueueProps) {
+    if ((props.message === undefined) === (props.messageObject === undefined)) {
```

Comment:
Perhaps a single required property of a union-like class instead of two optionals with runtime check?

[That seems like a repeating pattern between `string` and `Map<string => any>`. Maybe we can have a standard type for this?]

Reply:
I considered it, but I think it will be worse in the typical case. I'm expecting JSON object to be the common scenario and I'd hate to made that even harder to pass.

By the way, you're making the exact opposite argument in https://github.com/awslabs/aws-cdk/pull/2716. Which is it? :)

Reply:
I agree that this is not ideal, but The problem here is that it is a required argument, and having two optional arguments that are mutually exclusive results in poor IDE experience and we don’t have union types. 

Makes me wonder if perhaps we can implement automatic support for union types in Jsii so that we can use unions in typescript and other languages will be able to generate union-like classes....

Reply:
Yeeessss!!

Reply:
That would solve so much.

Reply:
How would you do it?

Reply:
I did it for consistency with Event Targets. That does mean we now have this abomination:

```ts
const task2 = new sfn.Task(this, 'Send2', {
    task: new tasks.SendToQueue(queue, {
        messageBody: TaskInput.fromObject({
            field1: 'somedata',
            field2: Data.stringAt('$.field2'),
        }),
        // Only for FIFO queues
        messageGroupId: '1234'
    })
});
```

Guess there's no getting around it... :x

Reply:
> How would you do it?

The only issue is how to automatically come up with a name for the union type. Anonymous won't be possible and any automatic mechanism will not be backwards compatible. Don't know if we get enough information from the TypeScript compiler, otherwise we could maybe force type unions to only be allowed if they're declared first:

```ts
export type Foo = Bar | Baz;
```

Then we can generate:

```
class Foo {
  static fromBar(Bar bar): Foo { }
  static fromBaz(Baz baz): Foo { }

  Bar getBar() { }
  Baz getBaz() { }
}
```

As an optimization, if the union type occurs in a props type for which we're already generating a builder, we could generate builder overloads like we do now:

```
class Builder {
  Builder withFoo(Foo foo) { }
  Builder withFoo(Bar foo) { }
  Builder withFoo(Baz foo) { }
}
```

Reply:
Yes, we will need to have an official name for the union type and also come up with a way to render the names of the various methods in case the union includes primitives/lists/maps...